### PR TITLE
PR: Add upsell notice.

### LINKED ISSUE #212: Use conditional Admin Notices to upsell add-ons
Consider using the following technique to upsell add-ons.

image

### PR DESCRIPTION
**Pull request by esudalaimadan**
Fix #212

### CODE REVIEW

File: email-log.php [hunk 1]
```diff
@@ -77,3 +77,5 @@ function email_log_deactivate() {
 
 // Fix compatibility issues with wpmandrill plugin.
 require_once plugin_dir_path( __FILE__ ) . 'include/compatibility/wpmandrill.php';
+
+require_once  __DIR__ . '/vendor/collizo4sky/persist-admin-notices-dismissal/persist-admin-notices-dismissal.php';
```
Comment (sudar) [inline email-log.php:81]: You don't need to include the file like this. You need to register the namespace. Check https://github.com/sudar/email-log/blob/dev/2.4.0/load-email-log.php#L29

File: include/Addon/License/Licenser.php [hunk 2]
```diff
@@ -68,6 +69,7 @@ public function load() {
 
 		add_action( 'el_before_addon_list', array( $this, 'render_bundle_license_form' ) );
 		add_action( 'el_before_logs_list_table', array( $this, 'render_more_fields_addon_upsell_message' ) );
+		add_action( 'el_before_logs_list_table', array( 'Upseller', 'render_auto_delete_logs_addon_upsell_message' ) );
```
Comment (sudar) [inline include/Addon/License/Licenser.php:72]: This should happen inside the `Upseller`.

Also move the Upsell message for more fields add-on to `Upseller`.

File: include/Addon/Upseller.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+defined( 'ABSPATH' ) || exit; // Exit if accessed directly.
+
+/**
+ * Handles upsell messages.
+ */
+
+ class Upseller{
+
+   	/**
```
Comment (sudar) [inline include/Addon/Upseller.php:11]: Add the `Loadie` interface to this class and then add the `load` method.

After that add it to the loadie list at https://github.com/sudar/email-log/blob/dev/2.4.0/load-email-log.php#L55

File: include/Addon/Upseller.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	 */
+	public function upsell_auto_delete_logs_in_settings_page() {
+		if ( $this->is_addon_active( 'Auto Delete Logs' ) ) {
+			return;
+		}
+
+		?>
+		<p>
+			<em>
+				<?php _e( 'You can also automatically delete logs if the database size increases using our <a href="https://wpemaillog.com/addons/auto-delete-logs/?utm_campaign=Upsell&utm_medium=wpadmin&utm_source=settings&utm_content=dl" target="_blank">Auto Delete Logs</a> add-on.', 'email-log' ); ?>
```
Comment (sudar) [inline include/Addon/Upseller.php:72]: Don't add links inside `_e` function. We have to use `printf`. I have fixed it in the other method. Use that as the sample code to fix this.

File: include/Core/UI/Setting/CoreSetting.php [hunk 5]
```diff
@@ -281,8 +281,10 @@ public function render_db_size_notification_settings( $args ) {
 					'<b>Save</b>'
 				); ?>
             </p>
-		<?php endif; ?>
 		<?php
+		endif;
+
+		do_action( 'el_after_db_size_notification_setting' );
```
Comment (sudar) [inline include/Core/UI/Setting/CoreSetting.php:287]: I have renamed the hook. Generally we should name the hook based on the place where it is invoked and not how the function that will use the hook.

Add the phpdoc block for this hook.

File: email-log.php (hunk 1 above)
Comment (esudalaimadan) [reply email-log.php]: Fixed.

File: include/Addon/License/Licenser.php (hunk 2 above)
Comment (esudalaimadan) [reply include/Addon/License/Licenser.php]: Fixed.  Moving More fields add-on upsell message is creating few problems. So more fields addon upsell message is not moved.

File: include/Addon/Upseller.php (hunk 3 above)
Comment (esudalaimadan) [reply include/Addon/Upseller.php]: Fixed.

File: include/Addon/Upseller.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	 */
+	public function load() {
+		add_action( 'el_before_logs_list_table', array( $this, 'auto_delete_logs_addon_upsell_message_loglist_page' ) );
+		add_action( 'el_auto_delete_logs_addon_upsell_message', array( $this, 'auto_delete_logs_addon_upsell_message_settings_page' ) );
+	}
+
+   	/**
+	 * Renders Upsell message for Auto delete logs add-on in Log list page.
+	 *
+	 * @since 2.4.0
```
[the remark this answers is not in the document]
Comment (esudalaimadan) [reply include/Addon/Upseller.php:28]: Fixed.

File: include/Addon/Upseller.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+	/**
+	 * Renders Upsell message for Auto delete logs add-on in Settings page.
+	 *
+	 * @since 2.4.0
+	 */
+	public function auto_delete_logs_addon_upsell_message_settings_page() {
+		?>
+		<div>
+			<p><?php _e( 'The Auto Delete Logs add-on allows you to automatically delete logs based on a schedule. <a href="https://wpemaillog.com/addons/auto-delete-logs/?utm_campaign=Upsell&utm_medium=wpadmin&utm_source=settings&utm_content=dl" target="_blank">Buy now</a>', 'email-log' ); ?></p>
```
[the remark this answers is not in the document]
Comment (esudalaimadan) [reply include/Addon/Upseller.php]: Fixed.

File: include/Addon/Upseller.php (hunk 4 above)
Comment (esudalaimadan) [reply include/Addon/Upseller.php:72]: Fixed.

File: include/Core/UI/Setting/CoreSetting.php (hunk 5 above)
Comment (esudalaimadan) [reply include/Core/UI/Setting/CoreSetting.php:287]: Fixed.

File: include/Addon/Upseller.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+		if ( $this->is_addon_active( 'Auto Delete Logs' ) ) {
+			return;
+		}
+
+		?>
+		<p>
+			<em>
+				<?php
+				printf(
+					__( 'You can also automatically delete logs if the database size increases using our <a href="https://wpemaillog.com/addons/auto-delete-logs/?utm_campaign=Upsell&utm_medium=wpadmin&utm_source=settings&utm_content=dl" target="_blank">Auto Delete Logs</a> add-on.', 'email-log' )
```
Comment (sudar) [inline include/Addon/Upseller.php:71]: You should not use html markup directly in `__` or `_e` functions. Check the above function to find out how to properly use `printf` function using format specifiers.

**Comment by sudar** [comment]
@esudalaimadan 

I left a small comment regarding `printf`.

Also did you also migrate the upsell message for More Fields add-on?

**Comment by esudalaimadan** [comment]
is_bundle_license_valid() method is not available in Upseller class, so I am unable to move upsell message for **more fields addon**.

**Comment by sudar** [comment]
@esudalaimadan 

Did you check the `is_addon_active` method in `Upseller`? We have access to `Licenser` and we just have to use that instance to call the `is_bundle_license_valid` method.

The `is_addon_active` is already doing that. You can even abstract that code into a new function.

**Pull request merged**